feat: Add Spotlight integration for local development debugging

DAcodedBEAT opened the pull request:
resolves: #846

Add SpotlightTransport that sends events to local Spotlight server for real-time debugging during development. The integration works by wrapping the existing transport and duplicating events to Spotlight.

Features:
- Spotlight transport decorator supporting custom URL overrides
- SENTRY_SPOTLIGHT environment variable for easy enabling
- Example program demonstrating usage patterns
- README documentation update

DAcodedBEAT commented:
> I'm ok merging this now instead of waiting for our refactoring, as the surface area is indeed rather small. If you could address the lint issue, we can review it and write docs prior the release.

Sorry for missing the lint errors! I've fixed it and re-pushed (trying to keep a clean git history for y'all). Let me know if there's anything else I can do to get this merged!

giortzisg reviewed (commented):
Thanks for the contribution. Left some small comments. Let's get them fixed and we can merge this.

DAcodedBEAT commented:
@giortzisg / @cleptric - may I please get some re-reviews on this?

giortzisg commented on transport.go:
```diff
@@ -746,6 +746,105 @@ func (noopTransport) FlushWithContext(context.Context) bool {
+	st.underlying.SendEvent(event)
+
+	// Always send to Spotlight
+	st.sendToSpotlight(event)
```
the send to spotlight call is sync here, so it kinda defeats the purpose of the async transport. We either need to spawn a go routine here or add it to the async queue of the transport.

giortzisg reviewed (commented):
We should update the implementation to also use the `internal/telemetry/scheduler` that we added for the new transport (we need to support both for now unfortunately) and also make the spotlight transport implementation async. Also left some comments for inconsistencies with the spec.

giortzisg commented on transport.go:
```diff
@@ -32,6 +32,7 @@ type Transport interface {
 	FlushWithContext(ctx context.Context) bool
 	Configure(options ClientOptions)
 	SendEvent(event *Event)
+	SendEnvelope(envelope *protocol.Envelope)
```
We should not implement sendEnvelope for the HttpTransport

giortzisg commented on transport.go:
```diff
@@ -507,6 +508,26 @@ func (t *HTTPTransport) SendEventWithContext(ctx context.Context, event *Event)
+// SendEnvelope sends an envelope to the Sentry server.
+// Currently converts envelope back to event for compatibility.
+// In the future, this will be the primary method.
+func (t *HTTPTransport) SendEnvelope(envelope *protocol.Envelope) {
```
same here.

giortzisg commented on transport.go:
```diff
@@ -777,6 +798,23 @@ func (t *HTTPSyncTransport) SendEvent(event *Event) {
 	t.SendEventWithContext(context.Background(), event)
 }
 
+func (t *HTTPSyncTransport) SendEnvelope(envelope *protocol.Envelope) {
```
and here.

giortzisg commented on transport.go:
```diff
@@ -887,6 +929,271 @@ func (noopTransport) FlushWithContext(context.Context) bool {
+	wg     sync.WaitGroup
+}
+
+func NewSpotlightTransport(underlying Transport) *SpotlightTransport {
```
following from the previous comment, when we setup the transport on the client, we already convert the `internal/http/transport` to a `Transport` interface with the `internalAsyncTransportAdapter` so there is no point in satisfying the `SendEnvelope` on the scope of this PR.

giortzisg commented on client.go:
```diff
@@ -1027,6 +1112,45 @@ func (client *Client) integrationAlreadyInstalled(name string) bool {
 
+// buildEnvelopeFromEvent builds an envelope from an event.
+// This is used to send events via the envelope-based transport API.
+func (client *Client) buildEnvelopeFromEvent(event *Event) *protocol.Envelope {
```
this is also not needed, the adapter should suffice.

giortzisg commented on client.go:
```diff
@@ -907,6 +986,12 @@ func (client *Client) processEvent(event *Event, hint *EventHint, scope EventMod
 			debuglog.Println("Event dropped: telemetry buffer full or unavailable")
 		}
 	} else {
+		// For Spotlight: build and send envelope for enhanced data collection
+		if client.options.Spotlight {
+			envelope := client.buildEnvelopeFromEvent(event)
+			client.Transport.SendEnvelope(envelope)
+		}
+		// Default path: send event directly for backwards compatibility
```
the transport is already wrapped with spotlight, so that part is not needed. the bugbot comments are valid.

giortzisg commented on transport.go:
```diff
@@ -887,6 +929,271 @@ func (noopTransport) FlushWithContext(context.Context) bool {
+	}
+}
+
+// Flush waits for the underlying transport to send pending events to Sentry.
+// In-flight Spotlight sends are best-effort and not waited on.
+func (st *SpotlightTransport) Flush(timeout time.Duration) bool {
+	return st.underlying.Flush(timeout)
+}
+
+func (st *SpotlightTransport) FlushWithContext(ctx context.Context) bool {
+	return st.underlying.FlushWithContext(ctx)
+}
```
this probably doesn't work with an empty dsn. the noop transport returns from flush immediately so no events would be delivered, we also need to wait for the spotlight events.

giortzisg commented on client.go:
```diff
@@ -365,6 +398,47 @@ func NewClient(options ClientOptions) (*Client, error) {
+
+	// Spotlight is a local development tool: always deliver 100% of events
+	// with full PII so the developer sees everything their app is generating.
+	if options.Spotlight {
+		if options.SampleRate != 1.0 {
+			debuglog.Printf("Overriding SampleRate from %.2f to 1.0 for Spotlight", options.SampleRate)
+			options.SampleRate = 1.0
+		}
+		if !options.SendDefaultPII {
+			debuglog.Println("Enabling SendDefaultPII for Spotlight")
+			options.SendDefaultPII = true
+		}
+	}
```
We shouldn't override it globally, because this would affect the sample rate and pii of the default sentry client

giortzisg commented:
Hey @DAcodedBEAT any updates on this?

DAcodedBEAT committed:
feat: add Spotlight integration for local development debugging

Adds support for Sentry Spotlight (https://spotlightjs.com/), a local
dev-time debugging sidecar. When enabled via ClientOptions.Spotlight or
the SENTRY_SPOTLIGHT env var, events are also forwarded to a local
Spotlight instance (default http://localhost:8969/stream).

- Uses the same transport path as real Sentry delivery by default.
- Never affects what's sent upstream: sample rate/PII overrides only
  apply with no DSN configured.
- Sidecar connectivity failures are backed off and logged independently.

DAcodedBEAT committed:
fix: two more Spotlight bugs from fresh bot review on PR #1085

- Envelope clone still raced with the transport: I'd made the clone
  itself synchronous earlier, but sendItem called transport.SendEnvelope
  (which can mutate the envelope on a background worker) before
  spotlight.Send read it. Swapped the order so Spotlight clones first.
- TracesSampler overrides TracesSampleRate, so forcing SampleRate to
  1.0 for Spotlight didn't help if a custom sampler was also set - it
  could still drop transactions. Now cleared it too.

DAcodedBEAT committed:
fix: same WaitGroup race in spotlightEnvelopeSender

Missed this when I fixed SpotlightTransport earlier. Send runs on the
scheduler's background goroutine and can race a concurrent Flush/Close
call from another goroutine - the pending counter only covers items
already buffered before Flush was called, not a brand new Send. This
one's actually worse than SpotlightTransport since it backs the
default Spotlight path everyone uses.

Switched to the same atomic counter + WaitForZero pattern, plus a
concurrent stress test under -race.

DAcodedBEAT committed:
fix: two more Spotlight bugs from cursor bot review

- Spotlight was silently skipped whenever the real Sentry transport
  was rate-limited or out of capacity, since processItems returned
  before ever reaching sendItem. Spotlight is a separate destination
  and should still get a copy in that case - otherwise local debugging
  misses exactly the events that got dropped on the way to Sentry.
  processItems now only gates the real transport send on rate-limit/
  capacity, not the Spotlight forward.
- SpotlightTransport.SendEvent handed the event pointer to a
  background goroutine that read it later (serializing to JSON),
  racing with anything mutating the event's nested maps after
  SendEvent returned. Moved the serialization/envelope-building
  synchronous, before the goroutine spawns - only the network POST
  runs async now, matching what spotlightEnvelopeSender already does.

DAcodedBEAT committed:
fix: pending counter bookkeeping bugs found by fable re-review

Two more instances of the counter-race pattern bots kept finding:

- Scheduler.Add incremented pending after Offer already made the item
  visible to run(), leaving a narrow window where a concurrent Flush
  could drain and decrement it before the increment landed. Increment
  now happens before Offer, rolled back if the item is rejected.
- The dropped-item callback decremented pending for every overflow
  reason, but "buffer_full_drop_newest"/"unknown_overflow_policy"
  reject the incoming item itself, which Add never incremented pending
  for - only "buffer_full_drop_oldest[_bucket]" evicts an
  already-counted item. Decrementing for the others drove pending
  permanently negative, so every Flush after that would spin to its
  deadline and report failure. Not reachable with the DropOldest
  policy this SDK actually configures, but real for any future policy
  change, so it's regression-tested directly against a DropNewest
  buffer.

DAcodedBEAT committed:
fix: two more bugs from fresh sentry bot review

- WaitForZero could report a timeout even when the counter had just
  hit zero, since select picks randomly between ready channels. Now
  double-checks the counter before returning false.
- Flush only force-drained buffers once, so anything captured right
  after (a log, say) sat waiting on its own batch timeout instead of
  being picked up, and Flush would spin to its deadline and fail. Not
  Spotlight-specific - this hit the plain flush path too. Now loops
  drain+wait until pending settles or ctx ends.

Two other bot comments on this diff were stale re-reports against an
earlier commit, already fixed since.

DAcodedBEAT committed:
fix: misleading no-DSN log on the default Spotlight path

noopTransport.Configure was already made Spotlight-aware for the
legacy transport path, but the default path (internal/http.NewAsyncTransport,
used by setupTelemetryProcessor) still logged "no events will be
delivered" for an empty/invalid DSN regardless. Misleading in the
recommended Spotlight-only setup (no DSN, per the example), where
events still reach Spotlight even though nothing goes to Sentry.
Threaded a Spotlight flag through TransportOptions purely to pick the
right wording.
